From: Jonathan Nieder <jrnieder@gmail.com>
Date: Tue, 29 Sep 2009 21:30:53 -0500
Subject: [PATCH 3/9] Add compress_cmd() for running gzip-like commands

By a strange coincidence (well, by design), gzip, bzip2, and lzma
use the same command-line syntax.  Factor out a function that can
be used to invoke any one of these commands.

This does not make the file any shorter (it barely breaks even).
It does add some peace of mind to the author: if any fix is
needed for the way all these commands are invoked, it only needs
to be applied to one place.  On the other hand, if some of these
commands start to require different options, the result could be
unpleasant.
---
 lib/dpkg/compression.h |   24 ++++++++++++------------
 1 files changed, 12 insertions(+), 12 deletions(-)

diff --git a/lib/dpkg/compression.h b/lib/dpkg/compression.h
index 7dd93c4..9205a6d 100644
--- a/lib/dpkg/compression.h
+++ b/lib/dpkg/compression.h
@@ -40,6 +40,15 @@ fd_fd_filter(int fd_in, int fd_out,
   ohshite(_("%s: failed to exec '%s %s'"), desc, cmd, args);
 }
 
+static void compress_cmd(int fd_in, int fd_out, const char *file,
+                         const char *cmd, char compression, const char *desc)
+{
+	char combuf[6];
+	strncpy(combuf, "-9c", sizeof(combuf));
+	combuf[1] = compression;
+	fd_fd_filter(fd_in, fd_out, file, cmd, combuf, desc);
+}
+
 static void decompress_noop(int fd_in, int fd_out, const char *desc)
 {
 	fd_fd_copy(fd_in, fd_out, -1, _("%s: decompression"), desc);
@@ -114,10 +123,7 @@ static void decompress_gzip(int fd_in, int fd_out, const char *desc)
 static void compress_gzip(int fd_in, int fd_out, char compression,
                           const char *desc)
 {
-  char combuf[6];
-  strncpy(combuf, "-9c", sizeof(combuf));
-  combuf[1]= compression;
-  fd_fd_filter(fd_in, fd_out, GZIP, "gzip", combuf, desc);
+	compress_cmd(fd_in, fd_out, GZIP, "gzip", compression, desc);
 }
 #endif
 
@@ -182,10 +188,7 @@ static void decompress_bzip2(int fd_in, int fd_out, const char *desc)
 static void compress_bzip2(int fd_in, int fd_out, char compression,
                            const char *desc)
 {
-  char combuf[6];
-  strncpy(combuf, "-9c", sizeof(combuf));
-  combuf[1]= compression;
-  fd_fd_filter(fd_in, fd_out, BZIP2, "bzip2", combuf, desc);
+	compress_cmd(fd_in, fd_out, BZIP2, "bzip2", compression, desc);
 }
 #endif
 
@@ -197,8 +200,5 @@ static void decompress_lzma(int fd_in, int fd_out, const char *desc)
 static void compress_lzma(int fd_in, int fd_out, char compression,
                           const char *desc)
 {
-	char combuf[6];
-	strncpy(combuf, "-9c", sizeof(combuf));
-	combuf[1] = compression;
-	fd_fd_filter(fd_in, fd_out, LZMA, "lzma", combuf, desc);
+	compress_cmd(fd_in, fd_out, LZMA, "lzma", compression, desc);
 }
-- 
1.6.5.rc1.199.g596ec

